### PR TITLE
Invoke expression tx

### PR DESCRIPTION
**Pull request by esuwu**

### CODE REVIEW

File: pkg/state/state.go [hunk 1]
```diff
@@ -984,7 +984,7 @@ func (s *stateManager) addNewBlock(block, parent *proto.Block, initialisation bo
 		height:         height,
 		initialisation: initialisation,
 	}
-	// Check and perform block's transactions, create balance diffs, write transactions to storage.
+	// Check and perform block's transactions, create balance diffs, write transactions to storage.\
```
Comment (nickeskov) [inline pkg/state/state.go:987]: Remove trailing backslash

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/proto/invoke_expression_transaction.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
+package proto
```
Comment (nickeskov) [inline pkg/proto/invoke_expression_transaction.go:1]: Delete this file and move content to `pkg/proto/transactions_with_proofs.go`.

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/ride/converters.go [hunk 3]
```diff
@@ -1142,36 +1173,57 @@ func convertArgument(arg proto.Argument) (rideType, error) {
...
+			for i, p := range transaction.Payments {
+				payments[i] = attachedPaymentToObject(p)
+			}
+			r["payments"] = payments
+		default:
+			r["payment"] = rideUnit{}
+			if len(transaction.Payments) > 0 {
+				r["payment"] = attachedPaymentToObject(transaction.Payments[0])
+			}
+		}
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/ride/converters.go:1204]: Actually, I did not. The only difference between version 5 and 6 is that in version 6 there are two fields `originCaller` and `originCaller`. And there is:
```
	if rideVersion >= 5 {
		r["originCaller"] = rideAddress(sender)
		r["originCallerPublicKey"] = callerPK
	}
```

File: pkg/proto/invoke_expression_transaction.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+func (tx InvokeExpressionTransactionWithProofs) GetFee() uint64 {
+	return tx.Fee
+}
+
+func (tx InvokeExpressionTransactionWithProofs) GetTimestamp() uint64 {
+	return tx.Timestamp
+}
+
+func (tx *InvokeExpressionTransactionWithProofs) Validate(scheme Scheme) (Transaction, error) {
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/proto/invoke_expression_transaction.go:59]: I don't think that this check `if tx.Type != InvokeExpressionTransaction` is relevant, as the type of transaction has been filled when initialized.

[the remark this answers is not in the document]
Comment (nickeskov) [reply pkg/proto/invoke_expression_transaction.go:59]: Yep, you're right. But I think second check is still relevant.

File: pkg/ride/converters.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickeskov) [reply pkg/ride/converters.go:1204]: I'm talking about payments, not `originCaller` and `originCallerPublicKey`.  🙂

File: pkg/state/invoke_applier.go [hunk 5]
```diff
@@ -792,7 +822,7 @@ func (ia *invokeApplier) applyInvokeScript(tx proto.Transaction, info *fallibleV
 	// Refuse payments to DApp itself since activation of BlockV5 (acceptFailed) and for DApps with StdLib V4.
 	disableSelfTransfers := info.acceptFailed && tree.LibVersion >= 4
 	if disableSelfTransfers && paymentsLength > 0 {
-		if sender == *scriptAddr {
+		if sender == *scriptAddr && !isInvokeExpression {
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/invoke_applier.go:825]: This is not only about `InvokeExpression`. In this function, InvokeScript is also handled.

[the remark this answers is not in the document]
Comment (nickeskov) [reply pkg/state/invoke_applier.go:825]: I'm talking only about `isInvokeExpression` flag. It's redundant.

**Review by nickeskov — APPROVED** [review]

**Pull request merged**